packages/cli/src/commands/store.js

```diff
@@ -8,11 +8,108 @@ import { E } from '@endo/far';
+
+  let offset = 0;
+  for (const array of arrays) {
+    result.set(array, offset);
+    offset += array.byteLength;
+  }
+
+  return result;
+};
+
+/**
+ * @param {AsyncIterable<Uint8Array>} reader
+ */
+const asyncConcat = async reader => {
+  const chunks = [];
+  for await (const chunk of reader) {
+    chunks.push(chunk);
+  }
+  return concat(chunks);
+};
+
+export const store = async ({
+  name,
+  agentNames,
+  storePath,
+  storeStdin,
+  storeText,
+  storeTextStdin,
+  storeJson,
+  storeJsonStdin,
+  storeBigInt,
+}) => {
+  const modes = {
+    storePath,
+    storeStdin,
+    storeText,
+    storeTextStdin,
+    storeJson,
+    storeJsonStdin,
+    storeBigInt,
+  };
+  const selectedModes = Object.entries(modes).filter(
+    ([_modeName, value]) => value !== undefined,
+  );
+  const selectedModeNames = selectedModes.map(([modeName]) => modeName);
+  if (selectedModes.length !== 1) {
+    // Usage error should be reported without trace.
+    // eslint-disable-next-line no-throw-literal
+    throw `Must provide exactly one store flag. Got flags for: (${selectedModeNames.join(
+      ', ',
+    )})`;
+  }
+
+  if (storePath !== undefined) {
+    const nodeReadStream = fs.createReadStream(storePath);
+    const reader = makeNodeReader(nodeReadStream);
+    const readerRef = makeReaderRef(reader);
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      await E(agent).storeBlob(readerRef, name);
+    });
+  } else if (storeStdin !== undefined) {
+    const reader = makeNodeReader(process.stdin);
+    const readerRef = makeReaderRef(reader);
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      await E(agent).storeBlob(readerRef, name);
+    });
+  } else if (storeText !== undefined) {
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      await E(agent).storeValue(storeText, name);
+    });
+  } else if (storeTextStdin !== undefined) {
+    const reader = makeNodeReader(process.stdin);
+    const bytes = await asyncConcat(reader);
+    const text = new TextDecoder().decode(bytes);
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      await E(agent).storeValue(text, name);
+    });
+  } else if (storeJson !== undefined) {
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      await E(agent).storeValue(JSON.parse(storeJson), name);
+    });
+  } else if (storeJsonStdin !== undefined) {
+    const reader = makeNodeReader(process.stdin);
+    const bytes = await asyncConcat(reader);
+    const text = new TextDecoder().decode(bytes);
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      await E(agent).storeValue(JSON.parse(text), name);
+    });
+  } else if (storeBigInt !== undefined) {
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      await E(agent).storeValue(BigInt(storeBigInt), name);
+    });
+  }
```

Comment:
Idk if this sparks joy for you, but:
```diff
diff --git a/packages/cli/src/commands/store.js b/packages/cli/src/commands/store.js
index be54e2240..b945c150e 100644
--- a/packages/cli/src/commands/store.js
+++ b/packages/cli/src/commands/store.js
@@ -8,6 +8,8 @@ import { E } from '@endo/far';
 
 import { withEndoAgent } from '../context.js';
 
+const { quote: q } = assert;
+
 /**
  * @param {Array<Uint8Array>} arrays
  * @returns {Uint8Array}
@@ -72,44 +74,40 @@ export const store = async ({
     )})`;
   }
 
+  const storeValueOrBlob = async (value, isBlob = false) => {
+    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
+      if (isBlob) {
+        await E(agent).storeBlob(value, name);
+      } else {
+        await E(agent).storeValue(value, name);
+      }
+    });
+  };
+
   if (storePath !== undefined) {
     const nodeReadStream = fs.createReadStream(storePath);
     const reader = makeNodeReader(nodeReadStream);
-    const readerRef = makeReaderRef(reader);
-    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
-      await E(agent).storeBlob(readerRef, name);
-    });
+    await storeValueOrBlob(makeReaderRef(reader), true);
   } else if (storeStdin !== undefined) {
     const reader = makeNodeReader(process.stdin);
-    const readerRef = makeReaderRef(reader);
-    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
-      await E(agent).storeBlob(readerRef, name);
-    });
-  } else if (storeText !== undefined) {
-    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
-      await E(agent).storeValue(storeText, name);
-    });
-  } else if (storeTextStdin !== undefined) {
-    const reader = makeNodeReader(process.stdin);
-    const bytes = await asyncConcat(reader);
-    const text = new TextDecoder().decode(bytes);
-    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
-      await E(agent).storeValue(text, name);
-    });
+    await storeValueOrBlob(makeReaderRef(reader), true);
+  } else if (storeBigInt !== undefined) {
+    await storeValueOrBlob(BigInt(storeBigInt));
   } else if (storeJson !== undefined) {
-    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
-      await E(agent).storeValue(JSON.parse(storeJson), name);
-    });
-  } else if (storeJsonStdin !== undefined) {
+    await storeValueOrBlob(JSON.parse(storeJson));
+  } else if (storeText !== undefined) {
+    await storeValueOrBlob(storeText);
+  } else {
     const reader = makeNodeReader(process.stdin);
     const bytes = await asyncConcat(reader);
-    const text = new TextDecoder().decode(bytes);
-    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
-      await E(agent).storeValue(JSON.parse(text), name);
-    });
-  } else if (storeBigInt !== undefined) {
-    await withEndoAgent(agentNames, { os, process }, async ({ agent }) => {
-      await E(agent).storeValue(BigInt(storeBigInt), name);
-    });
+    const rawText = new TextDecoder().decode(bytes);
+
+    if (storeJsonStdin !== undefined) {
+      await storeValueOrBlob(JSON.parse(rawText));
+    } else if (storeTextStdin) {
+      await storeValueOrBlob(rawText);
+    } else {
+      throw new Error(`Unrecognized mode: ${q(selectedModeNames[0])}`);
+    }
   }
 };

```

Reply:
I made it drier but did not desiccate. The `withEndoAgent` was obviously common to all paths and easy to wrap around the whole. Also was straightforward to rank the options by storeValue or storeBlob and in order of increasing ceremony within those bins, so it’ll probably read more easily, and still linearly.

Reply:
What you did is better 👍